### ISSUE TITLE
using custom modbus functions

### ISSUE THREAD

**Issue by ramonvogel1** [cabbi/modbus_client_serial#6]
Hi, i have a flutter project where i need to use the input and holding registers (which are working), as well as some custom modbus functions, and i'm having trouble implementing that. In my application this custom functions are like holding registers, what changes is just the function code. In python i was able to to that by just inherit the holding register model and change the function code directly. Can you perhaps give an example of how to use the custom functions? Thanks :)

**Comment by on443** [comment]
Give me some time, I'm thinking about a lean and clean solution for your case

**Comment by cabbi** [comment]
Give me some time, I'm thinking about a lean and clean solution for your case

**Comment by cabbi** [comment]
I've pushed a new 'custom_requests' branch in the 'modbus_client' package repository 
 (https://github.com/cabbi/modbus_client/tree/custom_requests). Please download it and play with the new 'custom_request_example.dart' example where you can change the read & write function codes.

Let me know if this fits your needs

```dart
import 'package:logging/logging.dart';
import 'package:modbus_client/modbus_client.dart';
import 'package:modbus_client_tcp/modbus_client_tcp.dart';

void main() async {
  // Simple modbus logging
  ModbusAppLogger(Level.FINE);

  // Create your own element type by defining your custom read & write function codes
  ModbusElementType myCustomType = ModbusElementType(
      ModbusFunctionCode(0x04, FunctionType.read), // ---> place here your custom read code
      ModbusFunctionCode(0x06, FunctionType.writeSingle), // ---> place here your custom write codes if supported
      ModbusFunctionCode(0x10, FunctionType.writeMultiple));

  // Create a modbus int16 register element
  var batteryTemperature = ModbusInt16Register(
      name: "BatteryTemperature",
      type: myCustomType,
      address: 22,
      uom: "°C",
      multiplier: 0.1,
      onUpdate: (self) => print(self));

  // Discover the Modbus server
  var serverIp = await ModbusClientTcp.discover("192.168.0.106");
  if (serverIp == null) {
    ModbusAppLogger.shout("No modbus server found!");
    return;
  }

  // Create the modbus client.
  var modbusClient = ModbusClientTcp(serverIp, unitId: 1);

  // Send a read request from the element
  await modbusClient.send(batteryTemperature.getReadRequest());

  // Ending here
  modbusClient.disconnect();
}

```

**Comment by ramonvogel1** [comment]
So as was trying to test this i've got an assignment error, in the send method, which was expecting a ModbusRequest type, not ModbusReadRequest or ModbusWriteRequest, i don't know if i did something wrong, i'm still new to dart, but here is the code i was using to test:
`import 'package:logging/logging.dart';
import 'package:modbus_client_serial/modbus_client_serial.dart';
import 'modbus_client_custom/modbus_client.dart';

void main() async {
  // Simple modbus logging
  ModbusAppLogger(Level.FINE);

  // Create your own element type by defining your custom read & write function codes
  ModbusElementType myCustomType = const ModbusElementType(
    ModbusFunctionCode(0x43, FunctionType.read),  
    ModbusFunctionCode(0x47, FunctionType.writeSingle),
  );

  // Create a modbus uint16 register element
  var genericRegister = ModbusUint16Register(
      name: "TestingRegister",
      type: myCustomType,
      address: 0,
      multiplier: 1,
      onUpdate: (self) => print(self));

  // Create the modbus client.
  var modbusClient = ModbusClientSerialRtu(portName: 'COM59', unitId: 2, baudRate: SerialBaudRate.b115200);

  await modbusClient.send(genericRegister.getReadRequest()); // --> ModbusReadRequest can't be assigned to ModbusRequest

  var req = genericRegister.getWriteRequest(2);
  var res = await modbusClient.send(req); // --> same error here
  print(res);

  // Ending here
  modbusClient.disconnect();
}
`

**Comment by cabbi** [comment]
Try to import 'package:modbus_client/modbus_client.dart' package:
``` dart
import 'package:logging/logging.dart';
import 'package:modbus_client/modbus_client.dart';
import 'package:modbus_client_serial/modbus_client_serial.dart';

void main() async {
   . . . 
   . . . 
   . . . 
```

**Comment by ramonvogel1** [comment]
If i just use this:
`import 'package:logging/logging.dart';
import 'package:modbus_client/modbus_client.dart';
import 'package:modbus_client_serial/modbus_client_serial.dart';
`

Then i get an error in this part:

`  ModbusElementType myCustomType = ModbusElementType(
    ModbusFunctionCode(0x43, FunctionType.read),  
    ModbusFunctionCode(0x47, FunctionType.writeSingle),
  );`

i think is because this way i'm not importing the branch with the FunctionType and FunctionCode editable, 'cause i'm getting this errors: 
"Undefined name 'FunctionType'. Try correcting the name to one that is defined, or defining the name" and 
"Generative enum constructors can only be used as targets of redirection. Try using an enum constant, or a factory constructor." on the ModbusFunctionCode.

Also, if i import the package and the custom_requests branch inside the modbus_client_custom folder i get an error for using the same class names and if i just import the folder with the branch i get the assignment error i was talking in my last comment

**Comment by cabbi** [comment]
You need to download the custom_event branch and work on that project/folder by editing the custom_request_example.dart file

**Comment by cabbi** [comment]
Please let me know if this version I've prepared for you is working well and as you expected.
In case I will release it on pub.dev.

**Comment by ramonvogel1** [comment]
it seems to be working fine, if anything happens i'll let you know :)

**Comment by cabbi** [comment]
Ok, I've published new versions 1.2.0 of each modbus_client packages, enjoy!
